### DISCUSSION TITLE
Extension Gallery Service

### DISCUSSION THREAD

**Discussion by CompuIves** [CodinGame/monaco-vscode-api#224] [Ideas]
I've tested the new remote VSCode server capability, and it's working really well which is super exciting!

Now that we support running server-side extensions, I wonder how feasible it would be to expose the extension gallery, allowing people to install extensions by themselves from a marketplace like [OpenVSX](https://open-vsx.org/). I'd be happy to contribute it, if you agree with the idea and if you can give some pointers on where to look!

**Comment by CGNonofr** [comment]
Of course, as long as it doesn't break compatibility with the current behavior, it can be another nice addition.

You may need to pull the vscode implementions of the IExtensionGalleryService and IExtensionManagementService services (instead of the fake implementations in missing-services.ts) as a new service override

**Reply by CompuIves** [reply]
Been working on this. I can pull `IExtensionGalleryService` and `IExtensionManagementService` and activate them, but the challenge is in showing the UI. To do this, I was initially thinking of importing `extensions.contribution.ts`, it registers the pane in the sidebar and the extension editor to view the extension (+ some commands). However, it looks like there's duplicated code the moment I import it.

I get an error, saying that some extensions already have been contributed, meaning that there's other code that has registered that specific contribution to the registry already. I'm getting this for e.g. the view registry.

<img width="699" alt="image" src="https://github.com/CodinGame/monaco-vscode-api/assets/587016/e08328a4-0d00-46b9-8825-d5d8d9bb2e83">

Is this something that you've seen before when overriding services? I see that the debug contribution is working, and that's also registering a new view, so there mist be something that can enable it 🤔 .

**Reply by CompuIves** [reply]
Ah! Got it to work by adding the service to the `optimizeDeps` field.

**Reply by CompuIves** [reply]
Getting further, I was able to load the extension gallery service, and now working on the contributions. One of the services `IExtensionWorkbenchService` requires `localizationService`. I've registered it, but even then it cannot find it, which is surprising...

<img width="933" alt="image" src="https://github.com/CodinGame/monaco-vscode-api/assets/587016/c597492e-723b-4413-a3f6-cd645b2f57a4">

The service is initialized [here](https://github.com/CodinGame/monaco-vscode-api/pull/238/files#diff-3de6b85a49d4ac75895dce844b1f885d9eb546a3e028ec7e702ad50c6de10c22R24), but even then the error still returns. It also complains about some other services. I wonder if there's a caching issue or something. Though when inspecting the service override object, they are there.

**Reply by CGNonofr** [reply]
`getServiceOverride` is not able to register new services, it can only override existing ones

you need to declare a fake minimal implementation in missing-services.ts

**Reply by CGNonofr** [reply]
but you probably don't need the real implementation at all, since only `setLocale` is used and what the vscode implementation does is probably not compatible with how we manage locales

**Reply by CompuIves** [reply]
Ahh, that explains a lot. I'll create those mock services 👍. Thanks for the help!

**Reply by CompuIves** [reply]
Tadaaaa 😄 

<img width="1433" alt="image" src="https://github.com/CodinGame/monaco-vscode-api/assets/587016/a4ece79e-e54f-4a24-af11-6ef030811014">

**Reply by CGNonofr** [reply]
is there something working except the empty panel? :smile:

**Reply by CompuIves** [reply]
<img width="1414" alt="image" src="https://github.com/CodinGame/monaco-vscode-api/assets/587016/b081940f-6e2e-4ee7-8ec6-af4494877aab">

Got it to work now! We can install extensions, both web and server! 🎉

**Reply by CompuIves** [reply]
Also search
<img width="1415" alt="image" src="https://github.com/CodinGame/monaco-vscode-api/assets/587016/8edcefde-6c4c-4c81-9acb-71e8eb1ce732">
